src/Microsoft.FeatureManagement/FeatureFilters/Recurrence/RecurrenceEvaluator.cs

```diff
@@ -0,0 +1,630 @@
+                    FindDailyPreviousOccurrence(time, settings, out previousOccurrence, out numberOfOccurrences);
+
+                    break;
+
+                case RecurrencePatternType.Weekly:
+                    FindWeeklyPreviousOccurrence(time, settings, out previousOccurrence, out numberOfOccurrences);
+
+                    break;
+
+                default:
+                    return false;
+            }
+
+            RecurrenceRange range = settings.Recurrence.Range;
+
+            if (range.Type == RecurrenceRangeType.EndDate)
+            {
+                return previousOccurrence <= range.EndDate;
+            }
+
+            if (range.Type == RecurrenceRangeType.Numbered)
+            {
+                return numberOfOccurrences <= range.NumberOfOccurrences;
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Find the closest previous recurrence occurrence before the provided timestamp according to the "Daily" recurrence pattern.
+        /// <param name="time">A timestamp.</param>
+        /// <param name="settings">The settings of time window filter.</param>
+        /// <param name="previousOccurrence">The closest previous occurrence.</param>
+        /// <param name="numberOfOccurrences">The number of occurrences between the time and the recurrence start.</param>
+        /// </summary>
+        private static void FindDailyPreviousOccurrence(DateTimeOffset time, TimeWindowFilterSettings settings, out DateTimeOffset previousOccurrence, out int numberOfOccurrences)
+        {
+            RecurrencePattern pattern = settings.Recurrence.Pattern;
+
+            DateTimeOffset start = settings.Start.Value;
+
+            Debug.Assert(time >= start);
+
+            int interval = pattern.Interval;
+
+            TimeSpan timeGap = time - start;
+
+            //
+            // netstandard2.0 does not support '/' operator for TimeSpan. After we stop supporting netstandard2.0, we can remove .TotalSeconds.
+            int numberOfInterval = (int) Math.Floor(timeGap.TotalSeconds / TimeSpan.FromDays(interval).TotalSeconds);
+
+            previousOccurrence = start.AddDays(numberOfInterval * interval);
+
+            numberOfOccurrences = numberOfInterval + 1;
+        }
+
+        /// <summary>
+        /// Find the closest previous recurrence occurrence before the provided timestamp according to the "Weekly" recurrence pattern.
+        /// <param name="time">A timestamp.</param>
+        /// <param name="settings">The settings of time window filter.</param>
+        /// <param name="previousOccurrence">The closest previous occurrence.</param>
+        /// <param name="numberOfOccurrences">The number of occurrences between the time and the recurrence start.</param>
+        /// </summary>
+        private static void FindWeeklyPreviousOccurrence(DateTimeOffset time, TimeWindowFilterSettings settings, out DateTimeOffset previousOccurrence, out int numberOfOccurrences)
```

Comment:
I'd like to chat about this algorithm, I'm wondering if we can do something a bit easier to follow.

In my mind a proposition is

```

1. first find day 0 (d0), it's the day representing the start day on the week of `Start`.
2. Find start day of most recent occurring week floor((targetDate - d0) / (interval * 7)) * (interval * 7)
3. if that's over 7 days ago, then previous occurence is the day with greatest offset of last occurring week
4. If gotten this far, then the current week has an occurence:
  i. if day offset > min offset then use maximum offset that is less than current to check if in window.
  ii.  If day offset < min offset,  then previous occurence is the day with greatest offset of last occurring week
```

but we can discuss